### PR TITLE
Feat/user verification login

### PR DESCRIPTION
**Pull request by johnnabil**
##### Why do we need this change?
  - Added functionality for verifying a new user using their email and a verification code sent to their email upon registration. This ensures that the user has access to the email they registered with
  - Users can now log in using their email and password with error handling. 
  - Upon successful login, the system generates a JWT (JSON Web Token) for the user. This token includes an expiration time.
  - Added a function to verify the validity of the access token. This helps to maintain user sessions and prevent unauthorized access.
  - Added a function to refresh the token, provided the refresh token is valid. This helps to ensure extending the user's session but in a secure way.
  - A logout function was implemented that ends the user's session by deleting the existing tokens.
  - Added the jsonwebtoken and cookie-parser packages to our project.

### CODE REVIEW

File: src/modules/User/services/userService.ts [hunk 1]
```diff
@@ -37,6 +38,95 @@ export class UserService {
   async findOne(args: Prisma.UserWhereInput) {
     return await this.userRepo.findOne({ email: args.email });
   }
+
+  //check verification code
```
Comment (kerolloz) [inline src/modules/User/services/userService.ts:42]: I think this comment serves no purpose

Comment (johnnabil) [reply src/modules/User/services/userService.ts:42]: it is just a clarity comment

File: src/modules/User/services/userService.ts [hunk 2]
```diff
@@ -37,6 +38,95 @@ export class UserService {
...
+    }
+    const isPasswordCorrect = await bcrypt.compare(password, user.password);
+    if (!isPasswordCorrect) {
+      throw new CustomError('Invalid Credentials', 401);
+    }
+    const accessToken = sign(
+      {
+        id: user.id,
+      },
+      'access_secret',
```
Comment (kerolloz) [inline src/modules/User/services/userService.ts:78]: No hard coded secrets, use env vars

Comment (johnnabil) [reply src/modules/User/services/userService.ts:78]: ok

File: src/modules/User/services/userService.ts [hunk 3]
```diff
@@ -37,6 +38,95 @@ export class UserService {
...
+    if (!user) {
+      throw new CustomError('User not found', 404);
+    }
+    const id = user.id;
+    if (user.isVerified) {
+      throw new CustomError('User already verified', 400);
+    }
+    if (user.verificationCode != args.verificationCode) {
+      throw new CustomError('Invalid verification code', 401);
+    }
```
Comment (kerolloz) [inline src/modules/User/services/userService.ts:54]: User not found shouldn't really be the response here in this case. This way you're exposing an endpoint that can be used to find out whether an email is registered in your db or not.

File: src/modules/User/services/userService.ts [hunk 4]
```diff
@@ -37,6 +38,95 @@ export class UserService {
...
+    } catch (error) {
+      throw new CustomError('Invalid or expired access token', 401);
+    }
+    if (!payload) {
+      throw new CustomError('Unauthenticated', 401);
+    }
+    const user = await this.userRepo.findOne({ id: payload.id });
+    if (!user) {
+      throw new CustomError('User not found', 404);
+    }
```
Comment (kerolloz) [inline src/modules/User/services/userService.ts:107]: I guess we can use a single error message in all these cases 

`Invalid token`, that's it. Don't expose too much info.

File: src/modules/User/services/userService.ts (hunk 1 above)
Comment (kerolloz) [reply src/modules/User/services/userService.ts:42]: It appears to be a bit redundant. Just a mirror of the function name.

**Review by kerolloz — APPROVED** [review]

**Review by mohamadelhinamy — DISMISSED** [review]

**Review by mohamadelhinamy — APPROVED** [review]

**Review by kerolloz — APPROVED** [review]

**Pull request merged**